### PR TITLE
FI-496 fix metadata errors

### PR DESCRIPTION
**Pull request by czh-orz**
This PR fixes some of the problems in the tests that were caused by metadata errors. It adds the 'ge', comparator and changes Procedure.occurenceDateTime to Procedure.performedDateTime.

**Submitter:**
- [X] This pull request describes why these changes were made
- [X] Internal ticket for this PR: https://oncprojectracking.healthit.gov/support/browse/FI-496
- [X] Internal ticket links to this PR
- [X] Internal ticket is properly labeled (Community/Program)
- [X] Internal ticket has a justification for its Community/Program label
- [X] Code diff has been reviewed for extraneous/missing code
- [ ] Tests are included and test edge cases
- [X] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: generator/uscore/metadata_extractor.rb [hunk 1]
```diff
@@ -284,6 +285,20 @@ def add_element_definitions(profile_definition, sequence)
         end
       end
 
+      def fix_metadata_errors(metadata)
+        # Procedure's date search param definition says Procedure.occurenceDateTime even though Procedure doesn't have an occurenceDateTime
+        procedure_sequence = metadata[:sequences].find { |sequence| sequence[:resource] == 'Procedure' }
+        procedure_sequence[:search_param_descriptions][:date][:path] = 'Procedure.performedDateTime'
```
Comment (Jammjammjamm) [inline generator/uscore/metadata_extractor.rb]: Does this also need to support `Procedure.performedPeriod`?

File: generator/uscore/metadata_extractor.rb [hunk 2]
```diff
@@ -180,7 +181,7 @@ def add_must_support_elements(profile_definition, sequence)
               sequence[:must_supports] <<
                 {
                   type: 'element',
-                  path: path.gsub('[x]', type['code'].slice(0).capitalize + type['code'].slice(1..-1))
+                  path: path.gsub('[x]', capitalize_first_letter(type['code']))
```
Comment (Jammjammjamm) [inline generator/uscore/metadata_extractor.rb]:
```suggestion
                  path: path.gsub('[x]', type['code'].capitalize)
```

File: generator/uscore/metadata_extractor.rb [hunk 3]
```diff
@@ -196,13 +197,10 @@ def add_must_support_elements(profile_definition, sequence)
...
-          if param.to_s != '_id'
-            path_parts[0] = sequence[:resource]
-            path = path_parts.join('.')
-          else
-            path = path_parts[0]
-          end
+          path = search_param_definition['expression']
+          path = path.gsub(/.where\((.*)/, '')
+          as_type = path.scan(/.as\((.*?)\)/).flatten.first
+          path = path.gsub(/.as\((.*?)\)/, capitalize_first_letter(as_type)) if as_type.present?
```
Comment (Jammjammjamm) [inline generator/uscore/metadata_extractor.rb]:
```suggestion
          path = path.gsub(/.as\((.*?)\)/, as_type.capitalize) if as_type.present?
```

File: generator/uscore/metadata_extractor.rb (hunk 2 above)
Comment (czh-orz) [reply generator/uscore/metadata_extractor.rb]: Can't do this because it makes the rest of string lowercase. dateTime becomes Datetime.

**Comment by arscan** [comment]
Looks like this needs to be rebased?

Also, I wonder if we should have some kind of assertion that the problems do actually exist before fixing them within that method.  I just wonder what happens if/when this capability statement gets updated and we have code in there modifying the structure.  Also in the case if we try to point this at another capability statement.

**Comment by czh-orz** [comment]
Yeah you have a point, but since 3.1.0 seems to be the last major update and we've stopped supporting older versions, then I thought this would be okay. Alternatively, I could go in and change the json resource files themselves.

**Comment by czh-orz** [comment]
Rebased and resolved codeclimate issue.

File: generator/uscore/metadata_extractor.rb [hunk 4]
```diff
@@ -320,6 +336,10 @@ def set_first_search(sequence, params)
         sequence[:searches].delete(search)
         sequence[:searches].unshift(search)
       end
+
+      def capitalize_first_letter(str)
```
Comment (radamson) [inline generator/uscore/metadata_extractor.rb]: Why is this needed over [`#capitalize`](https://ruby-doc.org/core-2.5.4/String.html#method-i-capitalize)?

File: lib/app/sequence_base.rb [hunk 5]
```diff
@@ -714,7 +714,7 @@ def can_resolve_path(element, path)
         path_ary = path.split('.')
         el_as_array = Array.wrap(element)
         cur_path_part = path_ary.shift.to_sym
-        return false if el_as_array.none? { |el| el.try(cur_path_part).present? }
+        return false if el_as_array.none? { |el| el.send(cur_path_part).present? }
```
Comment (radamson) [inline lib/app/sequence_base.rb]: why send over try?

File: lib/app/sequence_base.rb [hunk 6]
```diff
@@ -743,7 +743,7 @@ def resolve_element_from_path(element, path)
       def get_value_for_search_param(element)
         case element
         when FHIR::Period
-          element.start || element.end
+          'gt' + element.start || 'lt' + element.end
```
Comment (Jammjammjamm) [inline lib/app/sequence_base.rb]: You'll need an actual condition here because `'gt' + element.start` is always truthy.
